File: 07_fpga/c_golden/planta_bess.c
#include "planta_bess.h"

static double dmin(double a, double b) { return a < b ? a : b; }
static double dmax(double a, double b) { return a > b ? a : b; }

PlantaOut planta_bess(double SOC, double PBESS, double PBESS_ref,
                      double Pload, double Psource_max, int grid_status,
                      const EmsParam *param)
{
    PlantaOut o;
    double delta_SOC;

    PBESS_ref = dmin(PBESS_ref, param->PBESS_max);
    PBESS_ref = dmax(PBESS_ref, param->PBESS_min);

    o.PBESS_next = param->ab * PBESS + param->bb * PBESS_ref;
    o.PBESS_next = dmin(o.PBESS_next, param->PBESS_max);
    o.PBESS_next = dmax(o.PBESS_next, param->PBESS_min);

    if (SOC <= param->SOC_min && o.PBESS_next > 0.0)
        o.PBESS_next = 0.0;
    if (SOC >= param->SOC_max && o.PBESS_next < 0.0)
        o.PBESS_next = 0.0;

    if (grid_status == 1)
        o.Psource = Pload - o.PBESS_next;
    else
        o.Psource = 0.0;

    if (o.Psource > Psource_max)
        o.Psource = Psource_max;
    if (o.Psource < 0.0)
        o.Psource = 0.0;

    o.Pnao = Pload - o.Psource - o.PBESS_next;
    if (o.Pnao < 0.0)
        o.Pnao = 0.0;

    if (o.PBESS_next >= 0.0)
        delta_SOC = (o.PBESS_next * param->Ts / 3600.0) /
                    (param->Ebat * param->eta_descarga) * 100.0;
    else
        delta_SOC = (o.PBESS_next * param->Ts / 3600.0) *
                    param->eta_carga / param->Ebat * 100.0;

    o.SOC_next = SOC - delta_SOC;
    o.SOC_next = dmin(o.SOC_next, param->SOC_max);
    o.SOC_next = dmax(o.SOC_next, param->SOC_min);
    return o;
}

File: 01_matlab_base/planta_bess.m
function [SOC_next, PBESS_next, Psource, Pnao_atendida] = planta_bess(SOC, PBESS, PBESS_ref, Pload, Psource_max, grid_status, param)
% PLANTA_BESS
% Modelo simplificado da planta elétrica:
% data center + fonte principal + BESS/UPS.
%
% Convenção:
% PBESS > 0  -> bateria descarregando, fornecendo potência à carga
% PBESS < 0  -> bateria carregando, consumindo potência da fonte
%
% Entradas:
% SOC          -> estado de carga atual da bateria [%]
% PBESS        -> potência atual do BESS [kW]
% PBESS_ref    -> referência de potência do BESS [kW]
% Pload        -> potência da carga crítica [kW]
% Psource_max  -> limite disponível da fonte [kW]
% grid_status  -> 1 fonte disponível, 0 fonte indisponível
% param        -> estrutura de parâmetros
%
% Saídas:
% SOC_next       -> próximo SOC [%]
% PBESS_next     -> próxima potência real do BESS [kW]
% Psource        -> potência fornecida pela fonte [kW]
% Pnao_atendida  -> potência da carga não atendida [kW]

%% 1. Aplicar saturação na referência do BESS

PBESS_ref = min(PBESS_ref, param.PBESS_max);
PBESS_ref = max(PBESS_ref, param.PBESS_min);

%% 2. Dinâmica de primeira ordem do BESS/conversor

PBESS_next = param.ab * PBESS + param.bb * PBESS_ref;

%% 3. Aplicar saturação na potência real do BESS

PBESS_next = min(PBESS_next, param.PBESS_max);
PBESS_next = max(PBESS_next, param.PBESS_min);

%% 4. Proteção de SOC

% Se SOC está no mínimo, não deixa descarregar
if SOC <= param.SOC_min && PBESS_next > 0
    PBESS_next = 0;
end

% Se SOC está no máximo, não deixa carregar
if SOC >= param.SOC_max && PBESS_next < 0
    PBESS_next = 0;
end

%% 5. Calcular potência da fonte

% Balanço:
% Psource + PBESS = Pload
% Portanto:
% Psource = Pload - PBESS

if grid_status == 1
    Psource = Pload - PBESS_next;
else
    Psource = 0;
end

%% 6. Limitar potência da fonte

if Psource > Psource_max
    Psource = Psource_max;
end

if Psource < 0
    Psource = 0;
end

%% 7. Calcular potência não atendida

Pnao_atendida = Pload - Psource - PBESS_next;

if Pnao_atendida < 0
    Pnao_atendida = 0;
end

%% 8. Atualizar SOC

% Energia em kWh:
% P [kW] * Ts [s] / 3600 = kWh

if PBESS_next >= 0
    % Descarga
    delta_SOC = (PBESS_next * param.Ts / 3600) / (param.Ebat * param.eta_descarga) * 100;
else
    % Carga
    delta_SOC = (PBESS_next * param.Ts / 3600) * param.eta_carga / param.Ebat * 100;
end

% Como PBESS > 0 descarrega, SOC diminui
SOC_next = SOC - delta_SOC;

%% 9. Saturar SOC

SOC_next = min(SOC_next, param.SOC_max);
SOC_next = max(SOC_next, param.SOC_min);

end